### PR TITLE
Unify snarky interpreters

### LINKED ISSUE #136: Unify interpreters in snarky
Right now in snarky there are a bunch of different interpreters, which all essentially do the same recursion over the AST. I think these should be unified to reduce bug surface area

### PR DESCRIPTION
**Pull request by mrmr1993**
This PR reworks the interpreters `run_unchecked`, `run_and_check'`, `auxiliary_input` and `constraints_system` to all rely on the same utility function `run`. In particular:
* passing the `system` argument makes `run` create a constraints system using the given `RC1S_constraint_system.t`. If absent, all of these computations are skipped.
* the final argument of `run` is an `option`. If this is a `Some` value, `run` evaluates the `Checked.t`, otherwise these steps are skipped.

Checklist:

- [ ] Tests were added for the new behavior
- [ ] All tests pass (CI will check this if you didn't)
- [X] Does this close issues? List them:

closes #136

### CODE REVIEW

File: src/lib/snarky/src/snark0.ml [hunk 1]
```diff
@@ -915,159 +875,134 @@ module Make_basic (Backend : Backend_intf.S) = struct
...
+        match (x, s) with
+        | Some x, Some s ->
+            let s', y = As_prover.run x get_value s in
+            (Some s', Some y)
+        | _, _ -> (None, None)
       in
-      let system = R1CS_constraint_system.create () in
-      R1CS_constraint_system.set_primary_input_size system 0 ;
+      ignore
+        (Option.map system ~f:(fun system ->
```
Comment (imeckler) [inline src/lib/snarky/src/snark0.ml]: You can use `Option.iter`

File: src/lib/snarky/src/snark0.ml [hunk 2]
```diff
@@ -915,159 +875,134 @@ module Make_basic (Backend : Backend_intf.S) = struct
...
+          -> (a, s) t
+          -> Request.Handler.t
+          -> s option
+          -> s option * a =
        fun stack t handler s ->
         match t with
         | Pure x -> (s, x)
-        | With_constraint_system (f, k) -> f system ; go stack k handler s
+        | With_constraint_system (f, k) ->
+            ignore (Option.map ~f system) ;
```
Comment (imeckler) [inline src/lib/snarky/src/snark0.ml]: same (`Option.iter`)

File: src/lib/snarky/src/snark0.ml [hunk 3]
```diff
@@ -915,159 +875,134 @@ module Make_basic (Backend : Backend_intf.S) = struct
...
             go stack k handler s'
         | Add_constraint (c, t) ->
-            if not (Constraint.eval c get_value) then
-              failwithf "Constraint unsatisfied:\n%s\n%s\n"
-                (Constraint.annotation c)
-                (Constraint.stack_to_string stack)
-                () ;
-            Constraint.add ~stack c system ;
+            ignore
+              (Option.map system ~f:(fun system ->
```
Comment (imeckler) [inline src/lib/snarky/src/snark0.ml]: same (`Option.iter`)

File: src/lib/snarky/src/snark0.ml [hunk 4]
```diff
@@ -915,159 +875,134 @@ module Make_basic (Backend : Backend_intf.S) = struct
...
+                   (* NOTE: If s is None, we aren't evaluating and shouldn't consider constraints here. *)
+                   if Option.is_some s && not (Constraint.eval c get_value)
```
Comment (imeckler) [inline src/lib/snarky/src/snark0.ml]: Even in the case that we are evaluating, we still shouldn't necessarily do this check (evaluating a constraint is somewhat costly so we'd rather avoid it when we can). Maybe consider just passing a boolean that controls whether or not you do the check?

File: src/lib/snarky/src/snark0.ml [hunk 5]
```diff
@@ -915,159 +875,134 @@ module Make_basic (Backend : Backend_intf.S) = struct
...
+                       (Constraint.stack_to_string stack)
+                       () ;
+                   Constraint.add ~stack c system )) ;
             go stack t handler s
         | With_state (p, and_then, t_sub, k) ->
-            let s, s_sub = As_prover.run p get_value s in
+            let s, s_sub = run_as_prover (Some p) s in
             let s_sub, y = go stack t_sub handler s_sub in
-            let s, () = As_prover.run (and_then s_sub) get_value s in
+            let s, _ = run_as_prover (Option.map ~f:and_then s_sub) s in
```
Comment (imeckler) [inline src/lib/snarky/src/snark0.ml]: I think if this is a unit, better to just have the () pattern (it makes it more clear when reading that throwing this value away is fine).

File: src/lib/snarky/src/snark0.ml [hunk 6]
```diff
@@ -915,159 +875,134 @@ module Make_basic (Backend : Backend_intf.S) = struct
...
-            let var = Typ.Store.run (store value) store_field_elt in
-            let (), () = go stack (check var) handler () in
-            go stack (k {Handle.var; value= Some value}) handler s'
+        | Exists ({store; alloc; check; _}, p, k) -> (
+          match s with
+          | Some s ->
+              let s', value = Provider.run p get_value s handler in
+              let var = Typ.Store.run (store value) store_field_elt in
+              (* TODO: Push a label onto the stack here *)
+              let _, () = go stack (check var) handler (Some ()) in
```
Comment (imeckler) [inline src/lib/snarky/src/snark0.ml]: same () comment

File: src/lib/snarky/src/snark0.ml [hunk 7]
```diff
@@ -915,159 +875,134 @@ module Make_basic (Backend : Backend_intf.S) = struct
...
+          | None ->
+              let var = Typ.Alloc.run alloc alloc_var in
+              (* TODO: Push a label onto the stack here *)
+              let _, () = go stack (check var) handler None in
+              go stack (k {Handle.var; value= None}) handler None )
         | Next_auxiliary k -> go stack (k !next_auxiliary) handler s
       in
-      match go [] t0 Request.Handler.fail s0 with
+      let retval = go [] t0 Request.Handler.fail s0 in
+      ignore
```
Comment (imeckler) [inline src/lib/snarky/src/snark0.ml]: same (`Option.iter`)

File: src/lib/snarky/src/snark0.ml [hunk 8]
```diff
@@ -915,159 +875,134 @@ module Make_basic (Backend : Backend_intf.S) = struct
...
+             R1CS_constraint_system.set_auxiliary_input_size system
+               auxiliary_input_size )) ;
+      retval
+
+    (* TODO-someday: Add pass to unify variables which have an Equal constraint *)
+    let constraint_system ~num_inputs (t : (unit, 's) t) :
+        R1CS_constraint_system.t =
+      let input = Field.Vector.create () in
+      let next_auxiliary = ref (1 + num_inputs) in
+      let aux = Field.Vector.create () in
```
Comment (imeckler) [inline src/lib/snarky/src/snark0.ml]: It would be nice if we could avoid having to construct this vector when we don't use it (just from a code-cleanliness perspective) but not a huge deal

File: src/lib/snarky/src/snark0.ml (hunk 5 above)
Comment (mrmr1993) [reply src/lib/snarky/src/snark0.ml]: It's no longer a `unit`, it's a `unit option`. I've added a type signature instead, hopefully that makes it clearer.

File: src/lib/snarky/src/snark0.ml (hunk 8 above)
Comment (mrmr1993) [reply src/lib/snarky/src/snark0.ml]: I tried to do this (2c4dce470fed68e33ef5f768116ffbd9d5f76f5e), but it makes a bit of a mess of `run`. I haven't included it here, but I can pull it in/try a different approach if necessary.

File: src/lib/snarky/src/snark0.ml [hunk 9]
```diff
@@ -914,30 +916,28 @@ module Make_basic (Backend : Backend_intf.S) = struct
...
-              (Option.map system ~f:(fun system ->
-                   (* NOTE: If s is None, we aren't evaluating and shouldn't consider constraints here. *)
-                   if Option.is_some s && not (Constraint.eval c get_value)
-                   then
-                     failwithf "Constraint unsatisfied:\n%s\n%s\n"
-                       (Constraint.annotation c)
-                       (Constraint.stack_to_string stack)
-                       () ;
-                   Constraint.add ~stack c system )) ;
+            Option.iter system ~f:(fun system ->
```
Comment (imeckler) [inline src/lib/snarky/src/snark0.ml]: I think that checking constraints is orthogonal to whether or not we add them to the constraint system. E.g., in run_and_check we don't have a constraint system but we still need to be checking constraints.

Comment (mrmr1993) [reply src/lib/snarky/src/snark0.ml]: It should be adding the constraints either way (as long as there is a `system` to add them to). The `if eval_constraints && ... then` only covers the `failwithf`.
